bin/backup.sh

```diff
@@ -28,37 +31,42 @@ for i in {1..5}; do
 	echo "Dumping database cluster $i: $PGUSER@$PGHOST:$PGPORT"
 
 	# Wait for PostgreSQL to become available.
-	COUNT=0
+	count=0
 	until psql -l > /dev/null 2>&1; do
-		if [[ "$COUNT" == 0 ]]; then
+		if [[ "$count" == 0 ]]; then
 			echo "Waiting for PostgreSQL to become available..."
 		fi
-		(( COUNT += 1 ))
+		(( count += 1 ))
+		[ $count -lt $max_pg_wait_count ] || break
 		sleep 1
 	done
-	if (( COUNT > 0 )); then
-		echo "Waited $COUNT seconds."
+	if (( count > 0 )); then
+		echo "Waited $count seconds."
+		psql -l > /dev/null 2>&1 || {
+			echo "PostgreSQL still not available, trying next backup."
+			continue
+		}
 	fi
 
-	mkdir -p "/pg_dump"
+	mkdir "$work_area" || exit 1
```

Comment:
Can you elaborate on the problem with `-p`? If we do drop it, we don't need `|| exit 1` because we have `set -e` already.

Reply:
The issue with `-p` is twofold:
- it makes intermediate directories (that's its entire feature) - not an issue with a hardwired `/pg_dump` but definitely an issue with longer paths - with a longer path accidents in the longer path eg `/tmpp/work_area` do not get notice - badness just gets created in the filesystem.
- if the directory already exists, it is not an error - this means that 2 instances of the script can both "make" the work area, both try to use in concurrently, and both try to blow it away. All those things can variously lead to corrupt/truncated backups or removed-during-processing backups. All bad. `mkdir -p` in a script is usually a bug magnet. By not having `-p` we are just asking to make a _specific_ directory we want to be (a) new and (b) in a place where it is expected. `-p` discards these benefits.

Why might 2 scripts be running? If the backups take a very long time, something which creeps up on one for a fixed cronlike schedule. Forcing a mkdir is robust.

I like `|| exit` to make the flow control explicit, `-e` regardless. I use both `-e` and `-u` in my own scripts, but still make an `exit` like the above explicit. It aids readability. Also, that's a fun side effect of subshells, where `-e` effectively gets turned off and needs reenabling. Explicit flow control reduces the opportunity for this kind of misadventure.